### DISCUSSION TITLE
Simple search function for phrases

### DISCUSSION THREAD

**Discussion by erickpaquin** [autokey/autokey#727] [Ideas]
Hi guys,

Really love autokey, use it everyday. Saves my day all the time.

Because of this, I now have quite a lot of phrases that I use and don't always remember exactly how I named them etc.

Would be great then to have a simple search function to search for phrases. The search would return results if the key I'm searching for is part of either the phrase "name" or "abbreviation".

The search would be better if it was a "fuzzy" search but if not would still be useful.

Just a thought!

Thanks.

**Comment by josephj11** [comment]
This was discussed recently somewhere here or on Gitter. It's a good idea.

For now, since all all phrases and scripts are stored with plain text JSON metadata sidecar files, you can write a little bash code (or in any other language you prefer) to examine these files and produce whatever reports you like.

If you'd like help doing this, come over to [Gitter](https://gitter.im/autokey/autokey) to discuss it further.

You'd have to also look at the window filters for any triggers that are used multiple times.

**Reply by erickpaquin** [reply]
Good idea, thanks.

**Reply by josephj11** [reply]
Let us know what you come up with. I think others would find it useful.

Take a look at fzf which is great for fuzzy searches. I use it in several of my bash scripts.

**Comment by Elliria** [comment]
Caspx contributed a script to the wiki that you can find on the [https://github.com/autokey/autokey/wiki/Advanced-Scripts#searchPhrasesScripts](https://github.com/autokey/autokey/wiki/Advanced-Scripts#searchPhrasesScripts) page that opens a dialog that will search the titles of your phrases and scripts and returns a clickable list of matches so you can open the one you want from there. 

My copy of AutoKey has that script set to "Show in notification icon menu" so I don't have to hunt for it when I need it. I've got so many phrases and scripts that I'd be lost (or at least overworked) without it.

It doesn't search abbreviations, but it probably wouldn't be much work to make it do that, too. It also doesn't do fuzzy searches. I have no idea how much work it would be to get it to do that, but am guessing that might be a bit of a wrestle.

Either way, it's glorious as is and could also be a great start for what you're looking to do.

**Reply by josephj11** [reply]
Didn't know that was there. I'll have a look at it soon. Thanks.

**Reply by erickpaquin** [reply]
Thanks @Elliria. I'll have a look indeed!

**Comment by Elliria** [comment]
I seem to remember that that script used to open the selected script in the AutoKey editor, but I must be mistaken.

If you run that script and use the radio-list to choose a phrase or script from the results, its contents will be printed to the currently-active window and it won't be opened in the AutoKey editor.

Is there a way that that script can be modified to open the specified phrase or script in the AutoKey editor instead of printing its contents? I'm not seeing an API call for that.

**Reply by josephj11** [reply]
@Elliria  Still haven't tried that script. Really have to get around to it.

I think what you want would require at least a new API call.

However, with a little work, the script could be modified to use subprocess to open the selected action (script or phrase) in an external editor. You probably wouldn't want that to happen on every search result, so some sort of preview/confirmation would be good to add. That might require a bit more work.

I'm not sure exactly how that script works, but it looks like `f` contains the resulting file name. If that's the case and the rest of the path is good, then maybe copying it to the clipboard would do the trick. You could "pause" what you're doing in AutoKey and use the value in the clipboard as input to another program such as a text editor file selection dialog or command line argument.

**Comment by Elliria** [comment]
Ah, I didn't realize there wasn't already such a thing. At any rate, what the script currently does is return a radio list of all matching phrases and/or scripts so you can select one to have its contents printed to the currently-active window. You can't copy the script's name from the radio list, but just have to remember it and go open the script in AutoKey if that's what you were wanting.

My collection of phrases and scripts is so huge that I've used the script on and off for years now as a way to find out if I have a certain script in my collection and go to it, but that happens infrequently enough that I don't seem to remember that it doesn't take you to the chosen phrase/script.

I suppose it wouldn't be a bad idea to convert the script so that it puts its output into a dialog rather than printing it to the currently-active window, because if someone were to do what I did and run the script while AutoKey is open and a script is focused, it will print right into that script.

Here's my modification to just the try block in that script that offers four choices the user can pick from for how to display the result. The contents of the selection can be printed to the currently-active window or displayed in a dialog or the name of the selection can be printed to the currently-active window or displayed in a dialog. In this example, I've chose to display the name of the selection in a dialog:
```python
        try:
            with open(path) as f:
                phrase = f.read()
            """Enable the following line to print the contents of the selection to the currently-active window:"""
            # keyboard.send_keys(phrase)
            """Enable the following line to print the name of the selection to the currently-active window:"""
            # keyboard.send_keys(choice)
            """Enable the following line to display the contents of the selection in a dialog:"""
            # dialog.info_dialog(title='EXAMPLE', message=phrase)
            """Enable the following line to display the name of the selection in a dialog:"""
            dialog.info_dialog(title='EXAMPLE', message=choice)
```

**Reply by josephj11** [reply]
Took this over to Gitter so it doesn't clutter the issue.

**Reply by Elliria** [reply]
Okay, I followed you there.

**Comment by Yekutiel** [comment]
I suppose the following are probably not as elegant as you would like, but they might help you nonetheless.

If you want something on your desktop, [Recoll](https://www.lesbonscomptes.com/recoll/pages/index-recoll.html) is fast and powerful.

For a cloud solution perhaps you might upload your files to Google Drive, and then use Google Drive's search functionality. 

1. To restrict your search to a particular folder in Google Drive follow the three steps in the image below and then copy the URL. 
2. Then assign an AutoKey hotkey to that URL such as "Shift + F2" or "Ctrl + F9"
3. The next time you forget an AutoKey shortcut, open your browser, and run that shortcut in your browser's address bar. (That's a simple, slightly cumbersome approach).
4. Below is a more elegant approach that worked for me on Linux Mint, after I indicated to ChatGPT, "I want a python script that will open https://www.ibm.com/us-en in Google Chrome." and then indicated to ChatGPT "I am using Linux Mint. That script opened https://www.ibm.com/us-en in Firefox." 
5. Below, I substituted my Google Drive URL for https://www.ibm.com/us-en.
6. On Linux Mint, `pip3 install pyautogui` will install pyautogui.

```
import webbrowser, pyautogui

url = "https://drive.google.com/drive/search?q=parent:THE_UNIQUE_PORTION_OF_THE_URL"

# Specify the path to your Google Chrome executable on Linux Mint
chrome_path = "/usr/bin/google-chrome"

# Open the URL in Google Chrome
chrome_browser = webbrowser.get(chrome_path)
chrome_browser.open(url)  # or chrome_browser.open(url, new=2) to open in a new tab

# Sleep for 3 seconds
time.sleep(3)

# Simulate typing the forward slash key ("/")
# This will move the cursor to the search bar in Google Drive
pyautogui.press('/')
```

![Google Drive 2023-09-21_23-35](https://github.com/autokey/autokey/assets/16394972/b9605bc5-41d7-4d33-b0ee-4083d1554081)

**Comment by josephj11** [comment]
@Yekutiel recoll is a life saver. I don't use it very often, but it saves the day when I need it.

I'm fond of the Linux CLI and bash, so if I needed anything like this request, I would code it in bash and maybe use fzf for fuzzy searching. Once you use it for a while, grep can really help. The options to print lines before and after each match are nice if you need context.

**Reply by Yekutiel** [reply]
I played around with fzf for 5 to 10 hours a few years ago but I gave up on it because I didn't feel like going down yet another rabbit hole.

I spent around a hundred hours or so a few years ago helping to redesign Recoll's formerly terrible GUI a few years ago. The developer wasn't as stubborn as most about accepting my GUI suggestions. But I did have to bludgeon him with some very pointed logic to finally persuade him to implement many of my suggestions. 

I've learned to be nasty when dealing with developers... very nasty, because, well, ummmm they are generally arrogant ignoramuses who despise ordinary end users, and are generally unwilling to open their minds... unless they are pummeled mercilessly with irrefutable logic.

Ultimately I stopped using Recoll because Google Drive does everything I need... without the need for installing any software.

**Comment by josephj11** [comment]
While there is some truth to your assessment of developers, we're not all like that. I had some very good (quite unexpected) luck with a KDE developer not so long ago and I have had several other good experiences. But there are plenty out there that match your description. There is rather limited overlap between coding ability and interpersonal skills.

I prefer to do as much as possible on my local machines because I don't want anything unnecessary on the web for it to be mined or used for phishing... I like holding on to the illusion of privacy.

I just looked at one of my scripts which uses fzf. The command itself is pretty simple although I do a bit of sed on its output to get it the way I want to use it. I use fzf to search my activity log which is a simple text based journal and I use it in a script bound to Ctrl+f to search my terminal command history.

**Comment by Yekutiel** [comment]
> I just looked at one of my scripts which uses fzf. The command itself is pretty simple although I do a bit of sed on its output to get it the way I want to use it. I use fzf to search my activity log which is a simple text based journal and I use it in a script bound to Ctrl+f to search my terminal command history.

On average, a couple of dozen times per week I use the terminal to run all sorts of simple commands; but generally I try to use AutoKey instead.

fzf looks like it is the bees knees. However, I simply didn't want to go down that rabbit hole because I'm neither a developer, sys admin (er, uh, I mean DevOps engineer), nor do I want to become one. If someone made a nice GUI for fzf that I could wrap my head around, I'd be glad to use it. I tried a bunch of applications that were wrappers for fzf. Their convoluted "logic" was absurd.

Recoll actually worked very, very well for me. But I don't need it. Google Drive does what I need. In some ways (but not all ways) searching through my AutoKey scripts (Python scripts) on Google Drive actually works better than Recoll. I think that's because Google Drive uses some fuzzy logic that Recoll doesn't. But the Recoll dev did put some spiffy UI/UX features into Recoll which Google Drive is sorely missing.

I think Americans obsession with privacy are inane. Unless you are a terrorist, criminal, etc what good is all of this alleged privacy? And, of course, if some powerful official in the government wants to get you, he will. But does the chief of police, or some general in the US Army want to do you harm? I doubt it. 

That being said, all things being equal, I would prefer not to use clouds much at all because why depend on some company "out there", which might change their policies, go bankrupt, etc, when you can control stuff yourself? Nonetheless, generally, I'd rather use some clouds than deal with the headaches and hassles of not using clouds. 

I use Google Workspace (primarily Gmail, Google Docs, and Google Sheets) throughout my workdays. For $0 for the last 15 years or so, I have not had to deal with any of headaches and hassles like I used to in the "bad old days." I would gladly pay $10/month or even $25/month for Google Workspace, but Google keeps giving me what I need for free. I am not complaining!

I hated dealing with Windows, Mac, MS-DOS, and even, yeah, CP/M. Now I use Linux. I hate dealing with Linux too... but, I hate it less than any other operating system I've ever dealt with. Linux Mint 21.2 Cinnamon is far from delightful. I really do not like it. However, it's much less hassle operating system than I've ever dealt with. 

A decade ago I was hoping I'd be able to simply use a Chromebook, but, alas, I want/need a bunch of features Chromebooks don't offer.

Also, the security ordinary versions of Linux offers is very, very reassuring. (Yes, I know Chromebooks run on Linux, yet throngs of nerds outside of Google aren't able to review it as they can with ordinary versions of Linux). See, although privacy per se doesn't worry me; hackers do worry me. They definitely worry me because they can steal my money and my identity.

Sure I could run Thunderbird, LibreOffice Writer, and LibreOffice Calc, but then I would have way, way, way more headaches and hassles. Been there, done that... for many, many years. It's kind of the whole "Cattle not Pets" argument. [The History of Pets vs Cattle and How to Use the Analogy Properly](https://cloudscaling.com/blog/cloud-computing/the-history-of-pets-vs-cattle/). Playing with computers isn't a hobby for me; it's a hassle.

> There is rather limited overlap between coding ability and interpersonal skills.

That's a polite way of phrasing it.

I unapologetically attack arrogant (and concomitantly ignorant) engineers with a ferocity that actually surprises myself.

The fact that AutoKey lacks a simple, built-in method for searching is an excellent example of why I generally despise engineers. How difficult would it be to have such a feature? Sure, searching millions of records quickly is difficult. But I would guess the average user of AutoKey has no more than a few hundred scripts. Therefore, an ordinary, old, simple, slow search application would suffice for most users' needs.

For example, I tried to get the developer of Recoll to work with the developer of CopyQ (a very powerful clipboard, with a bad GUI which the dev refused to improve, despite my helpful suggestions) so that CopyQ would be able to use Recoll's search functionality. Neither dev was interested in working with the other. They enjoy being the Lord of the Manor. They are generally arrogant jerks who despise end users.

I've worked with precisely two excellent engineers in my life who, believe it or not, actually happened to be very nice, very humble guys. The rest of the engineers I have worked with have been bad and/or arrogant. When dealing with engineers my policy is simple: they are guilty until proven innocent.

**Reply by josephj11** [reply]
Then I hope you don't consider me an engineer!

Keep in mind that AutoKey was written by one developer who had his hands full fixing bugs and adding basic features. He eventually burned out and left the project. Since then, we have had a few wonderful developers work on the project, but never anything like "full time". Under these conditions it's really great that the project keeps working at all!

I have several enhancement requests that I really want that I may never get. It's not because of any shortcomings of the developers. They just don't have time to do them.

I was pretty fond of CP/M. It didn't do much, but it was simple to use. I used BASIC, C, and Z80 assembler on it, as well as some apps. One set of BASIC programs I wrote each started with if 1/1 !=1 then stop (don't remember the exact syntax). LOL. Sometimes it ran out of memory or something and that test would actually trigger. I had 64k ram and almost never used it all. It was a huge upgrade when I got a 10MB disk drive!

It wouldn't  be that hard to call a bash script running fzf from within AutoKey or to turn the bash script into a standalone GUI program using yad or Zenity.

The only tricky part is interacting with fzf to select the desired search result. If I were doing this, that would be the first thing to figure out.

I have scripts like that and even launch them from desktop or panel icons. Once you see how it works, most of the code is pretty simple. My scripts tend to be a bit larger because I add defensive programming to them to check for valid arguments, etc, and to give me really good error messages. I can share almost all of them if there's anything you're interested in looking at.

**Reply by Yekutiel** [reply]
Engineers, especially FOSS developers, despite claims to the contrary, tend to focus on adding cool features which are often irrelevant to ordinary users; whereas, they typically ignore mundane features which ordinary users would find useful.

I presume most of AutoKey's features are irrelevant to ordinary users. All I need AutoKey for is it's GUI which allows me to easily create and edit Abbreviations and Hotkeys.

Like I indicated previously, Google Drive works for my AutoKey searching needs.

Despite its popularity, AutoKey is a typical, barely maintained, mediocre FOSS project written by a dev who didn't care much about what ordinary users need.

I presume fzf is utterly, totally, and completely unnecessary for ordinary user's AutoKey searching needs. A simple 50 Python script would probably suffice to enable users to search for items they had created in AutoKey. But engineers tend to enjoy prattling on about all sorts of cool, glittery, inane solutions that will never be implemented. Don't they?

**Reply by josephj11** [reply]
I strongly disagree with you on that. It does have a few very arcane features like visgrep() and using scripts to create other scripts and phrases, but most of it is simple phrases (that were specifically designed for ease of use by non-programmers) and a pretty basic API for scripting. Some of the modules we use like the keyboard module have some strange quirks, but we didn't write that and it was (and probably still is) the best that was available at the time.

If you look through all our issues (and those from the original branch), you will see a large number of examples of our developers working extensively with users to solve problems and add features. We have had very little "prattling" on here and almost no arrogance. Everyone is treated with respect.

Over  the last few years, there has been a major effort to build out and enhance our wiki for the benefit of all our users.

I still believe that AutoKey is the best desktop automation solution for Linux. There are several newer competitors, but none seem to have all our features yet. Espanso looks like the strongest contender, but I haven't actually used it.

No one was suggesting building an API or other interface between AutoKey and fzf. It's just a nice tool that can be used for searching and since all of AutoKey's actions are coded as plain text files, fzf could readily be applied to them.

Insulting a project is not likely to elicit support for any fixes or enhancements that you might want. In general, programming is a lot harder than non-coders believe it to be, especially for applications that need to interact with other applications and desktop environments that were definitely not designed with that in mind.

**Reply by Yekutiel** [reply]
It seems to me that the strongest contender to AutoKey is ChatGPT helping folks write Python scripts. I suppose AI is going to render AutoKey useless within the next 5 to 10 years.

I used to make all sorts of polite, patient, and courteous suggestions for features on AutoKey, yet none of them were ever implemented. Not one. Not ever. 

Instead of prattling on about fzf, why don't you simply put up information on the AutoKey wiki about using Recoll with AutoKey and to my Google Drive search instructions. Also, why not include a note about them in a phrase that is preloaded in AutoKey? Both solutions both work very well for searching through AutoKey scripts, yet, obviously, the overwhelming majority of AutoKey users don't know about using them to search through their AutoKey scripts.

I mean, you do care about helping users. Don't you? You aren't an arrogant, self-centered, ignorant engineer who enjoys prattling on about all sorts of tech that will never be implemented, yet ignores explaining mundane features that most users would find helpful. Right?

**Reply by josephj11** [reply]
You, sir, are being a troll which is as bad or worse than any claimed arrogance you feel you have experienced here.

Any user with a GitHub account  can add content to our wiki. It should be done in a way that fits in with our indexed headings and must be written in a way that you would want it explained to you - politely and clearly. Since you seem to be an expert recoll user, you could probably write a better explanation that I can.

I have spent several thousand hours answering user questions for over a decade often answering the same question patiently many times - and have added a number of articles to the wiki. If you do not already know that, then you haven't bothered to look.  I recently deleted a couple of my old test scripts (on my local notebook) that were named with your name from when I was helping you research or solve an issue.

This is a purely volunteer project. I don't get paid to do this. Neither does anyone else in this project. If you feel you can do something more or better than we have done, make a contribution to the project by adding to the wiki or improving the code itself. Casting aspersions serves no constructive purpose.

Since I am not an AutoKey developer, I would not have been the one to implement any of your requests unless they involved documentation.

Our developers as volunteers work on whatever they are interested in. We have no control over them other than by discussing issues and presenting arguments. Luckily, we have had several very good ones who have done a lot to improve AutoKey's internals and overall robustness rather than just adding features that they wanted for themselves.
